ScriptRunner respects -d under -save

som-snytt opened the pull request:
Under -save, use -d target.jar.

If -d is a directory, then assemble the usual script.jar.

Also, `settings.d` is too minimalistic; prefer `settings.outdir`.

Also improve the idiom for "default" `settings.outdir` so it is not always `isSetByUser`.

Fixes scala/bug#2943

som-snytt committed:
Settings outdir is canonical

som-snytt committed:
ScriptRunner respects -d under -save

Under -save, use -d target.jar.

If -d is a directory, then assemble
the usual script.jar.

Correct main selection.

If no module is found, scripting should wrap always.

lrytz commented on src/compiler/scala/tools/nsc/ScriptRunner.scala:
```diff
@@ -83,57 +87,69 @@ abstract class AbstractScriptRunner(settings: GenericRunnerSettings) extends Scr
-        if (scriptFile endsWith ".jar") scriptFile
-        else scriptFile.stripSuffix(".scala") + ".jar"
+      val outputToJar = settings.outdir.value.endsWith(".jar")
+      def stripped = List(".scala", ".sc").find(scriptFile.endsWith).map(scriptFile.stripSuffix).getOrElse(scriptFile)
```
Is there anything special about `.sc` files? Adding support for makes them more "official".

som-snytt replied:
I happened to have `hello.sc` in my snippets dir. I see that if it stripped any extension starting with `.sc`, it would also cover `.script`. There is no automatic special handling at this time.

som-snytt replied:
`.sc` is also worksheet file in doti IDE.

lrytz reviewed (commented):
Looks good, some test failures

som-snytt commented:
In summary, `-save` uses `-d` to let the compiler output the jar, which is fixed now.

For debug, can `-save -d mydir` and it will leave classes in mydir and also assemble the jar. Not entirely sure that's useful.

The previous behavior is that `-save empty.scala` or `-e ''` will not fail, but it will also not succeed trivially. If there is no main to run, no jar is produced under `-save`, so it will always compile. That ought to be improved.

som-snytt committed:
BooleanSetting takes a default

Remove withDefault.

OutputSetting is StringSetting loosely coupled
to outputDirs.

MultiStringSetting takes a default

Back out clearSetByUser, which worked around initialization
of the setting.

som-snytt commented:
Normally, I am pleased when it's not my fault:
```
[error] java.io.IOException: Access to URL https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots;build.timestamp=1588892487296/org/scala-lang/scala-dist/2.13.3-bin-f6f83ab-SNAPSHOT/scala-dist-2.13.3-bin-f6f83ab-SNAPSHOT.pom was refused by the server: Forbidden; Response Body: {
[error]   "errors" : [ {
[error]     "status" : 403,
[error]     "message" : "'scala-ci' is blocked due to recurrent login failures, please try again in 3 seconds"
[error]   } ]
```
That is quite a forbidding message.

som-snytt commented:
/rebuild

som-snytt commented:
/synch

som-snytt commented:
Unrelated `presentation/doc` test was failing on commit to suppress warn on another test. Removed the spuriously failing commit.

SethTisue reviewed (approved):
@som-snytt this looks ready for merge. mind squashing the 1st commit out of existence, and combining the 4th and 5th into one? (or if that seems like annoying busywork to you, just hit "merge"?)

som-snytt commented:
OK, I'll rebase and squash. There is a semantic conflict at that first commit which I will de-conflict.

dwijnand commented on src/compiler/scala/tools/nsc/ScriptRunner.scala:
```diff
@@ -83,57 +87,69 @@ abstract class AbstractScriptRunner(settings: GenericRunnerSettings) extends Scr
+          val reg = new CloseableRegistry
+          ZipAndJarClassPathFactory.create(fil, settings, reg)
+        }
+        else DirectoryClassPath(location.jfile)
```
`ClassPathFactory.newClassPath`, and it'll even create the pesky `CloseableRegistry` for ya!

dwijnand commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -215,16 +216,16 @@ class MutableSettings(val errorFn: String => Unit, val pathFactory: PathFactory)
   def MultiChoiceSetting[E <: MultiChoiceEnumeration](name: String, helpArg: String, descr: String, domain: E, default: Option[List[String]] = None) =
     add(new MultiChoiceSetting[E](name, helpArg, descr, domain, default))
-  def OutputSetting(outputDirs: OutputDirs, default: String) = add(new OutputSetting(outputDirs, default))
+  def OutputSetting(outputDirs: OutputDirs, default: String) = add(new OutputSetting(default).tap(_ => outputDirs.setSingleOutput(default)))
```
Kind of sucks to have to manually apply the postSetHook here, especially with the hook defined so far away :-/

som-snytt replied:
This is initial value, not post-set, so I found it clear even on second reading. It crossed my mind to wonder why the setting and the outputDirs are two things.

dwijnand replied:
```scala
  val outdir        = OutputSetting (outputDirs, ".").withPostSetHook(s => try outputDirs.setSingleOutput(s.value) catch { case FatalError(msg) => errorFn(msg) })
```

My point is it's the same logic (modulo try/catch) repeated.

Maybe, seeing as there's only one `OutputSetting`, it would be possible to kill this wrapping def and call new/tap/withPostSetHook at `val outdir`?

som-snytt replied:
Yup. And yup.

dwijnand commented on test/junit/scala/tools/nsc/ScriptRunnerTest.scala:
```diff
@@ -13,10 +13,16 @@ class ScriptRunnerTest {
+    ScriptRunner(s).runScriptText("", Nil) match {
+      case Some(e) => throw e
+      case None    => ()
+    }
```
`foreach(throw _)`

som-snytt replied:
pair program next time over zoom.

dwijnand reviewed (approved):
Non-blocking suggestions.  Good to find ways to get rid of `clearSetByUser` and `withDefault`.

som-snytt commented:
Something broke when I squashed. :(
